File: src/main/java/insogroup/utils/utils/IOC/strategies/IOCStrategiesWithGenerators/StandardCacheParamsIOCStrategy.java
package insogroup.utils.utils.IOC.strategies.IOCStrategiesWithGenerators;

import insogroup.utils.utils.IOC.IOC;
import insogroup.utils.utils.IOC.exceptions.ResolveStrategyException;
import insogroup.utils.utils.IOC.interfaces.IOCStrategy;
import org.json.JSONObject;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class StandardCacheParamsIOCStrategy implements IOCStrategy {
    protected Map<String, JSONObject> templates = new HashMap<>();

    protected IOCStrategy generator;
    /**
     * @param params
     * Await: <pre>
     *     {
     *         "targetGenerator" : "ioc name or canonical classpath",
     *         "generatorParams" : object,
     *         "templates" : {
     *
     *         }
     *     }
     * </pre>
     */
    public StandardCacheParamsIOCStrategy(JSONObject params) throws ResolveStrategyException {
        String iocName = params.getString("targetGenerator");
        if (params.has("generatorParams")) {
            generator = IOC.resolve(iocName, params.get("generatorParams"));
        } else {
            generator = IOC.resolve(iocName);
        }

        if (params.has("templates")) {
            JSONObject templates = params.getJSONObject("templates");

            for (String key : templates.keySet()) {
                templates.put(key, templates.getJSONObject(key));
            }
        }
    }

    /**
     * Await at least one argument - name of dep
     * @param args Args for using in strategy
     * @return
     * @throws ResolveStrategyException
     */
    @Override
    public Object getInstance(Object... args) throws ResolveStrategyException {

        try {
            if (args.length > 0) {
                String name = (String) args[0];
                if (templates.containsKey(name)) {
                    return generator.getInstance(templates.get(name));
                } else {
                    JSONObject params;
                    if (args.length > 1) {
                        params = (JSONObject) args[1];
                    } else {
                        params = new JSONObject();
                    }
                    return generator.getInstance(params);
                }
            }
            return generator.getInstance(new JSONObject());
        } catch (ResolveStrategyException e) {
            throw new ResolveStrategyException("Can't generate object with params: " + Arrays.toString(args), e);
        }
    }
}

File: src/main/kotlin/com/github/insanusmokrassar/utils/IOC/strategies/IOCStrategiesWithGenerators/StandardCacheParamsIOCStrategy.kt
package com.github.insanusmokrassar.utils.IOC.strategies.IOCStrategiesWithGenerators

import com.github.insanusmokrassar.iobjectk.interfaces.IObject
import com.github.insanusmokrassar.iobjectk.interfaces.has
import com.github.insanusmokrassar.utils.IOC.IOC
import com.github.insanusmokrassar.utils.IOC.exceptions.ResolveStrategyException
import com.github.insanusmokrassar.utils.IOC.interfaces.IOCStrategy
import com.github.insanusmokrassar.utils.IOCField
import com.github.insanusmokrassar.utils.configField
import com.github.insanusmokrassar.utils.presetField
import com.github.insanusmokrassar.utils.strategyField
import java.util.*

class StandardCacheParamsIOCStrategy : IOCStrategy {
    protected var templates: Map<String, IObject<Any>> = HashMap()

    protected var generator: IOCStrategy

    /**
     * @param params
     * Await:
     * <pre>
     *     {
     *         "strategy" : "ioc name or canonical classpath",
     *         "config" : object,
     *         "preset" : {
     *
     *         }
     *     }
     * </pre>
     */
    @Throws(ResolveStrategyException::class)
    constructor(params: IObject<Any>) {
        val IOC = params.get<IOC>(IOCField)
        val strategyName = params.get<String>(strategyField)
        if (params.keys().contains(configField)) {
            generator = IOC.resolve(strategyName, params.get(configField))
        } else {
            generator = IOC.resolve<IOCStrategy>(strategyName)
        }

        if (params.has(presetField)) {
            val templates = params.get<IObject<Any>>(presetField)

            for (key in templates.keys()) {
                templates.put(key, templates.get<IObject<Any>>(key))
            }
        }
    }

    /**
     * Await at least one argument - name of dep
     * @param args Args for using in strategy
     * *
     * @return
     * *
     * @throws ResolveStrategyException
     */
    @Throws(ResolveStrategyException::class)
    override fun <T> getInstance(vararg args: Any): T {

        try {
            if (args.isNotEmpty()) {
                val name = args[0] as String
                if (templates.containsKey(name)) {
                    return generator.getInstance(templates[name]!!)
                } else {
                    val params: Array<Any>?
                    if (args.size > 1) {
                        params = Arrays.copyOfRange(args, 1, args.size)
                    } else {
                        params = null
                    }
                    if (params == null) {
                        return generator.getInstance()
                    } else {
                        return generator.getInstance(params)
                    }
                }
            }
            return generator.getInstance()
        } catch (e: ResolveStrategyException) {
            throw ResolveStrategyException("Can't generate object with params: $args", e)
        }

    }
}
